### PR TITLE
parser: type name resolution preamble

### PR DESCRIPTION
**Pull request by sploiselle**
To support resolving types as catalog objects, we need to shift a few things in the parser, namely moving map key checks out, and disallowing timelike keywords as column aliases.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5078)

### CODE REVIEW

File: demo/billing/src/mz.rs [hunk 1]
```diff
@@ -195,7 +195,7 @@ fn billing_agg_view(unit: &str) -> String {
     format!("CREATE MATERIALIZED VIEW billing_agg_by_{unit} AS
 SELECT date_trunc('{unit}', interval_start::timestamp) AS {unit}, client_id, meter, cpu_num, memory_gb, disk_gb, sum(value)
 FROM billing_records
-GROUP BY {unit}, client_id, meter, cpu_num, memory_gb, disk_gb",
+GROUP BY \"{unit}\", client_id, meter, cpu_num, memory_gb, disk_gb",
```
Comment (benesch) [inline demo/billing/src/mz.rs:198]: As alluded to previously the fact that you had to introduce quoting here is worrisome to me! There could very well be user views out there that use months as column names and somehow PostgreSQL is able to support that.

Restricting date/time fields for use in aliases is also backwards incompatible of course, but that one is unavoidable, it seems! We should get a release note mentioning that though.

Comment (benesch) [reply demo/billing/src/mz.rs]: This should be fixed up now. Details are in the amended commit message.

**Review by sploiselle — COMMENTED** [review]
@benesch tysm for finding that very subtle bug in the parser I had papered over with those quotes; this LGTM

**Comment by benesch** [comment]
Thanks for identifying the issue with time units as aliases in the first place!

**Pull request merged**